internal/cmd/dev.go

```diff
@@ -89,16 +85,350 @@ func (s *statement) UnmarshalJSON(data []byte) error {
+	NamedArgs *[]hranaNamedArg `json:"named_args,omitempty"`
+	WantRows  *bool            `json:"want_rows"`
+}
+
+type hranaCond struct {
+	Type  string       `json:"type"`
+	Step  *int         `json:"step,omitempty"`
+	Cond  *hranaCond   `json:"cond,omitempty"`
+	Conds *[]hranaCond `json:"conds,omitempty"`
+}
+
+type hranaBatchStep struct {
+	Stmt *hranaStmt `json:"stmt,omitempty"`
+	Cond *hranaCond `json:"condition,omitempty"`
+}
+
+type hranaBatch struct {
+	Steps []hranaBatchStep `json:"steps"`
+}
+
+type hranaExecuteReq struct {
+	Stmt hranaStmt `json:"stmt"`
+}
+
+type hranaBatchReq struct {
+	Batch hranaBatch `json:"batch"`
+}
+
+type hranaStmtResult struct {
+	Cols     []hranaCol     `json:"cols"`
+	Rows     [][]hranaValue `json:"rows"`
+	Affected int32          `json:"affected_row_count"`
+	LastIR   *string        `json:"last_insert_rowid"`
+}
+
+type hranaSuccessResponse struct {
+	Result hranaStmtResult `json:"result"`
+}
+
+type hranaBatchResponse struct {
+	Results []*hranaStmtResult `json:"step_results"`
+	Errors  []*hranaError      `json:"step_errors"`
+}
+
+type hranaBatchSuccessResponse struct {
+	Result hranaBatchResponse `json:"result"`
+}
+
+func executeHranaStatement(stmt hranaStmt, db *sql.DB) (*hranaStmtResult, *hranaError) {
+	var result hranaStmtResult
+
+	statement := *stmt.Sql
+	params := make([]interface{}, 0)
+
+	if stmt.Args != nil {
+		for _, narg := range *stmt.Args {
+			params = append(params, narg.Value)
+		}
+	}
+
+	if stmt.NamedArgs != nil {
+		for _, narg := range *stmt.NamedArgs {
+			params = append(params, sql.Named(narg.Name, narg.Value.Value))
+		}
+	}
+
+	if verboseFlag {
+		fmt.Printf("Executing %s, %s\n", internal.Emph(statement), internal.Emph(params))
+	}
+
+	rows, err := db.Query(statement, params...)
+	if err != nil {
+		// FIXME: I can't get the actual SQLite error code from this
+		return nil, hranaErr(err)
+	}
+	defer rows.Close()
+
+	result.Cols = make([]hranaCol, 0)
+	columns, _ := rows.Columns()
+	if columns == nil {
+		columns = []string{}
 	}
+
+	for _, col := range columns {
+		// Consistent with the response from SQLd
+		if strings.ToLower(col) == "null" {
+			col = "NULL"
+		}
+		c := hranaCol{col}
+		result.Cols = append(result.Cols, c)
+	}
+
+	result.Rows = make([][]hranaValue, 0)
+	for rows.Next() {
+		values := make([]interface{}, len(columns))
+		pointers := make([]interface{}, len(columns))
+		for i := range values {
+			pointers[i] = &values[i]
+		}
+		err := rows.Scan(pointers...)
+		if err != nil {
+			return nil, hranaErr(err)
+		}
+		row := make([]hranaValue, 0)
+		for _, val := range values {
+			switch v := val.(type) {
+			case nil:
+				row = append(row, hranaValue{Type: "null"})
+			case int64:
+				row = append(row, hranaValue{Type: "integer", Value: fmt.Sprintf("%d", v)})
+			case float64:
+				row = append(row, hranaValue{Type: "float", Value: v})
+			case []byte:
+				row = append(row, hranaValue{Type: "blob", Value: string(v)})
```

Comment:
Note that the correct format is `{"type": "blob", "base64": base64-encoded blob}`